src/sentry/tasks/summaries/weekly_reports.py

```diff
@@ -211,76 +224,127 @@ def prepare_organization_report(
 
     if not report_is_available:
         logger.info(
-            "prepare_organization_report.skipping_empty", extra={"organization": organization_id}
+            "prepare_organization_report.skipping_empty",
+            extra={"batch_id": batch_id, "organization": organization_id},
         )
         return
 
     # Finally, deliver the reports
+    batch = OrganizationReportBatch(ctx, batch_id, dry_run, target_user, email_override)
     with sentry_sdk.start_span(op="weekly_reports.deliver_reports"):
-        logger.info("weekly_reports.deliver_reports", extra={"organization": organization_id})
-        deliver_reports(
-            ctx, dry_run=dry_run, target_user=target_user, email_override=email_override
+        logger.info(
+            "weekly_reports.deliver_reports",
+            extra={"batch_id": batch_id, "organization": organization_id},
         )
+        batch.deliver_reports()
 
 
-def deliver_reports(
-    ctx: OrganizationReportContext,
-    dry_run: bool = False,
-    target_user: User | None = None,
-    email_override: str | None = None,
-) -> None:
-    """
-    For all users in the organization, we generate the template context for the user, and send the email.
-    """
-    # Specify a sentry user to send this email.
-    template_context: Mapping[str, Any] | None = None
-    user_id: int | None = None
-
-    if email_override:
-        target_user_id = (
-            target_user.id if target_user else None
-        )  # if None, generates report for a user with access to all projects
-        user_template_context_by_user_id_list = prepare_template_context(
-            ctx=ctx, user_ids=[target_user_id]
-        )
-        if user_template_context_by_user_id_list:
-            user_template_context_by_user_id = user_template_context_by_user_id_list[0]
-            template_context = user_template_context_by_user_id.get("context")
-            user_id = user_template_context_by_user_id.get("user_id")
-            if template_context and user_id:
-                send_email(
-                    ctx=ctx,
-                    template_ctx=template_context,
-                    user_id=user_id,
-                    dry_run=dry_run,
-                    email_override=email_override,
+@dataclass(frozen=True)
+class OrganizationReportBatch:
+    ctx: OrganizationReportContext
+    batch_id: uuid.UUID
+
+    dry_run: bool = False
+    target_user: User | None = None
+    email_override: str | None = None
+
+    def _prevent_duplicate_delivery(self) -> None:
+        from sentry.utils import redis  # TODO: Use sentry.buffer.redis instead?
+
+        cluster = redis.clusters.get("default").get_local_client_for_key("weekly_reports")
+        key = f"{self.batch_id}:{self.ctx.organization.id}"
```

Comment:
I can try to dig through the data from 2 weeks ago, but I'm curious if we know if this happens to an org, does every user get duplicate sends? Or should we also store the user id to filter to a potentially smaller list of user ids to send to?

Reply:
Yes, we have a few options as to where in the flow we want to call this. We could definitely store a key per user, provided that won't take up too much space in Redis.

Reply:
I can try to pull how many users it would be but I imagine it's quite a large number 😬 Do you know how to find out the limits of what we can put in Redis?

Reply:
For posterity I pulled this and it's around 3m users, we verified over Slack that Redis will handle this just fine. 

I made a [Redash query](https://redash.getsentry.net/queries/6899/source) to validate that we don't necessarily see the same number of duplicates across an organization (and cross referenced `mail.queued` logs and checked the counts per user for the weekly report email) so I feel good about us using the user id in the Redis key as well.